azurelinuxagent/ga/update.py

```diff
@@ -16,266 +16,399 @@
 
-"""
-Handles self update logic
-"""
+
+def get_update_handler():
+    return UpdateHandler()
+
+
 class UpdateHandler(object):
 
     def __init__(self):
         self.osutil = get_osutil()
         self.protocol_util = get_protocol_util()
+        self.running = True
         self.last_etag = None
+        self.last_attempt_time = None
         self.agents = []
-        self.error_record = {}
 
-    def run(self):
+    def run_latest(self):
         """
-        1. If self-update is enabled, check for new versions.
-           otherwise, use current agent
-        2. Invoke run-exthandlers task
+        This method is called from the daemon to find and launch the most
+        current, downloaded agent.
         """
-        self.mk_ga_dir()
-
-        self.load_error_record() 
-
-        # Check if the GoalState incarnation has updated
-        updated = self.check_for_update()
-        if not updated:
-            return 
-
-        # Assume using the current agent (which None causes to be used)
-        latest_agent = None
-
-        # TODO: Enable autoupdate -- force disable for now
-        # if conf.get_autoupdate_enabled():
-        #     logger.info("Auto update enabled")
-        #     latest_agent = self.get_latest_agent()
-
-        try:
-            self.run_extensions(latest_agent)
-        except UpdateError as e:
-            add_event(u"WALA", is_success=False, message=ustr(e))
-            if latest_agent is not None:
-                latest_agent.err.mark_failure()
-        
-        self.save_error_record()
-    
-    def mk_ga_dir(self):
-        dir_path = os.path.join(conf.get_lib_dir(), GA_DIR)
-        fileutil.mkdir(dir_path, mode=0o700)
-
-    def get_latest_agent(self):
-        available_agent = [agent for agent in self.agents if not agent.err.is_blacklisted()]
- 
-        return available_agent[0] if len(available_agent) >= 1 else None
-
-    def run_extensions(self, latest_agent):
-        agent_bin = sys.argv[0]
-
-        if latest_agent is not None:
-            logger.info(u"The latest guest agent version is : {0}", 
-                        latest_agent.version)
+        latest_agent = self.get_latest_agent()
+        if latest_agent == None:
+            agent_bin = sys.argv[0]
+            agent_name = "'Installed'"
+        else:
             agent_bin = latest_agent.get_agent_bin()
-
-            if not latest_agent.is_downloaded():
-                try:
-                    latest_agent.download()
-                except Exception as e:
-                    raise UpdateError(u"Download failed", e)
+            agent_name = latest_agent.name()
 
         devnull = open(os.devnull, 'w')
         try:
             child = subprocess.Popen([agent_bin, 'run-exthandlers'], 
                                       stdout=devnull, stderr=devnull)
+            ret = child.wait()
+            if ret == None:
+                ret = 1
+            if ret != 0:
```

Comment:
@hglkrijger Good question. I had not considered the impact on scvmm. Could you fill me in what I need to be aware of?